File: ACM-ICPC/OnlineJudge/vjudge/contest/6th/B_.py
n, limit = map(int, input().split())
data = list(map(int, input().split()))

data.sort()
amount = sum(data)

L, R = 0, 0
if amount % n == 0:
    L = R = int(amount / n)
else:
    L = int(amount / n)
    R = L + 1
l, r = R, 2 * 10 ** 9
while l <= r:
    mid = int((l + r) / 2)
    cnt = 0
    for it in data[::-1]:
        if it > mid:
            cnt += it - mid
        else:
            break

    if cnt > limit:
        l = mid + 1
    elif cnt <= limit:
        R = mid
        r = mid - 1

    cnt = 0

l, r = 0, L
while l <= r:
    mid = int((l + r) / 2)
    cnt = 0
    for it in data:
        if it < mid:
            cnt += mid - it
        else:
            break

    if cnt <= limit:
        L = mid
        l = mid + 1
    elif cnt > limit:
        r = mid - 1

    cnt = 0

print(str(R - L))

File: ACM-ICPC/OnlineJudge/vjudge/contest/6th/B.cc
#include <iostream>
#include <cstdio>
#include <algorithm>
#include <climits>

using std::cin;
using std::cout;
using std::endl;

int dat[500010];
int main() {
    int n, k, sum = 0;
    cin >> n >> k;
    for (int i = 0; i < n; i++) {
        scanf("%d", dat + i);
        sum += dat[i];
    }
    std::sort(dat, dat + n);
    int L, R;
    if (sum % n == 0) {
        L = R = sum / n;
    } else {
        L = sum / n;
        R = L + 1;
    }
    int l = R, r = INT_MAX;
    while (l <= r) {
        long long mid = (l + r) / 2;
        int cnt = 0;
        for (int i = n - 1; i >= 0; i--) {
            if (dat[i] > mid) {
                cnt += dat[i] - mid;
            } else {
                break;
            }
        }
        if (cnt > k) {
            l = mid + 1;
        } else {
            R = mid;
            r = mid - 1;
        }
    }
    l = 0, r = L;
    while (l <= r) {
        long long mid = (r + l) / 2;
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            if (dat[i] < mid) {
                cnt += mid - dat[i];
            } else {
                break;
            }
            if (cnt <= k) {
                L = mid;
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
    }

    cout << R - L << endl;
    return 0;
}
